android-kotlin/app/src/main/java/com/mongodb/app/LoginActivity.kt

```diff
@@ -0,0 +1,125 @@
+        accountExistsToggleButton.setOnClickListener { toggleExistingUser() }
+
+        loginButton.isEnabled = true
+        accountExistsToggleButton.isEnabled = true
+    }
+
+    override fun onBackPressed() {
+        // Disable going back to the MainActivity
+        moveTaskToBack(true)
+    }
+
+    private fun onLoginSuccess() {
+        // successful login ends this activity, bringing the user back to the project activity
+        startActivity(Intent(this, ItemsActivity::class.java))
+        // finish()
+    }
+
+    private fun onLoginFailed(errorMsg: String) {
+        Log.e(TAG(), errorMsg)
+        Toast.makeText(baseContext, errorMsg, Toast.LENGTH_LONG).show()
+    }
+
+    // Function to validate that the username adheres to all username rules.
+    private fun validateUsername(username: String): Boolean {
+        if (username.isEmpty()) {
+            return false // An empty username is not valid.
+        }
+        return true
+    }
+
+    // Function to validate that the password adheres to all password rules.
+    private fun validatePassword(password: String): Boolean {
+        if (password.isEmpty()) {
+            return false // An empty password is not valid.
+        }
+        return true
+    }
+
+    // Function to swap the log in use case with the create account use case.
+    private fun toggleExistingUser() {
+        this.createUser = !createUser
+        if (createUser) {
+            loginButton.setText(getString(R.string.create_account))
+            accountExistsToggleButton.setText(getString(R.string.already_have_account))
+        } else {
+            loginButton.setText(getString(R.string.log_in))
+            accountExistsToggleButton.setText(getString(R.string.does_not_have_account))
+        }
+    }
+
+    // handle user authentication (login) and account creation
+    private fun login() {
+        // while this operation completes, disable the button to login or create a new account
+        loginButton.isEnabled = false
+
+        val username = this.username.text.toString()
+        val password = this.password.text.toString()
+
+        if (!validateUsername(username)) {
+            onLoginFailed("Invalid username")
+            return
+        }
+        if (!validatePassword(password)) {
+            onLoginFailed("Invalid password")
+            return
+        }
+
+        if (createUser) {
+            // register a user using the Realm App
+            myApp.emailPassword.registerUserAsync(username, password) {
+                // re-enable the buttons after user registration returns a result
+                loginButton.isEnabled = true
+                if (!it.isSuccess) {
+                    onLoginFailed( "Error: ${it.error}")
+                } else {
+                    Log.i(TAG(), "Successfully registered user.")
+                    // when the account has been created successfully, log in to the account
+                    this.createUser = false
```

Comment:
Changing the mode and calling "login" again doesn't feel very direct and intentional. Ideally, you could rename this `login()` method to something like `onLoginButtonClicked`, which is more descriptive of what it does. Then put `logIn()` and `registerUser()` in their own methods, so that you can call `logIn()` directly from the success handler of `registerUser()`. This would clean up the code by reducing the indentation level and also make it more intentional.

Reply:
I agree, I've cleaned up a lot of the functions to be more clear